Add mid-cycle 1.30 blog article (#45411)
kubernetes/website: content/en/blog/_posts/2024-03-08-mid-cycle-1.30.md

Reviewer: Localization teams will also thank us if the Markdown source is wrapped at around 100 glyphs per line.

Author: > Localization teams will also thank us if the Markdown source is wrapped at around 100 glyphs per line.

I have no idea what this means. Would you please explain it to me?

Author: Sentence case done for headings and KEP titles. Will fix Markdown wrap after other content changes are done.

parameters less opaque. Instead of handling the semantics of all claim parameters themselves,
drivers could manage resources and describe them using a specific "structured model" pre-defined by
Kubernetes. This would allow components aware of this "structured model" to make decisions about
these resources without outsourcing them to some third-party controller. For example, the scheduler
would be able to allocate claims rapidly, without back-and-forth communication with dynamic resource
allocation drivers. Work done for this release centers on defining the framework necessary to enable
different "structured models" and to implement the "named resources" model. This model allows
listing individual resource instances and, compared to the traditional device plugin API, adds the
ability to select those instances individually via attributes.

### Node memory swap support ([KEP-2400](https://kep.k8s.io/2400))

In Kubernetes v1.30, memory swap support on Linux nodes gets a big change to how it works - with a
strong emphasis on improving system stability. In previous Kubernetes versions, the `NodeSwap`

> feature gate was disabled by default and, when enabled, it used `UnlimitedSwap` behavior as the

Reviewer 2: ```suggestion
feature gate was disabled by default, and when enabled, it used `UnlimitedSwap` behavior as the
```

default behavior. To achieve better stability, `UnlimitedSwap` behavior (which might compromise node
stability) will be removed in v1.30.

> The updated, still-beta support for swap on Linux nodes will be available by default, although the

Reviewer 2: ```suggestion
The updated, still-beta support for swap on Linux nodes will be available by default. However, the
```

default behavior will be to run the node set to `NoSwap` (not `UnlimitedSwap`) mode. In `NoSwap`
mode, the kubelet supports running on a node where swap space is active, but Pods don't use any of
the page file. You'll still need to set `--fail-swap-on=false` for the kubelet to run on that node.
However, the big change is the other mode: `LimitedSwap`. In this mode, the kubelet actually uses
the page file on that node and allows Pods to have some of their virtual memory paged out.
Containers (and their parent pods)  do not have access to swap beyond their memory limit, but the

> system can still make use of the swap space if available.

Reviewer 2: ```suggestion
system can still use the swap space if available.
```

Kubernetes' Node special interest group (SIG Node) will also update the documentation to help you
understand how to use the revised implementation, based on feedback from end users, contributors,
and the wider Kubernetes community.

> For more details on Linux node swap support in Kubernetes, read the previous [blog
> post](/blog/2023/08/24/swap-linux-beta/) or the [node swap
> documentation](/docs/concepts/architecture/nodes/#swap-memory).

Reviewer 2: ```suggestion
Read the previous [blog post](/blog/2023/08/24/swap-linux-beta/) or the [node swap
documentation](/docs/concepts/architecture/nodes/#swap-memory) for more details on 
Linux node swap support in Kubernetes.
```

### Support user namespaces in pods ([KEP-127](https://kep.k8s.io/127))

> [User namespaces](/docs/concepts/workloads/pods/user-namespaces) is a Linux-only feature to better
> isolate pods to prevent or mitigate several CVEs rated high/critical, including the just published

Reviewer 2: ```suggestion
[User namespaces](/docs/concepts/workloads/pods/user-namespaces) is a Linux-only feature that better
isolates pods to prevent or mitigate several CVEs rated high/critical, including the just-published
```

[CVE-2024-21626](https://github.com/opencontainers/runc/security/advisories/GHSA-xr7r-f8xq-vfvv). In
Kubernetes 1.30, support for user namespaces is migrating to beta, and now supports pods with and
without volumes, custom UID/GID ranges and more!

### Structured authorization configuration ([KEP-3221](https://kep.k8s.io/3221))

The Structured Authorization Configuration feature is moving to beta and will be enabled by default.
This
[feature](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3221-structured-authorization-configuration)
enables the creation of authorization chains with multiple webhooks with well-defined parameters
that validate requests in a particular order and allows fine-grained control – such as explicit Deny
on failures. The configuration file approach even allows you to specify
[CEL](/docs/reference/using-api/cel/) rules to pre-filter requests before they are dispatched to
webhooks, helping you to prevent unnecessary invocations. The API server also automatically reloads
the authorizer chain when the configuration file is modified.

You must specify the path to that authorization configuration using the `--authorization-config`

> command line argument. If you want to keep on using command line flags instead of using a

Reviewer 2: ```suggestion
command line argument. If you want to keep using command line flags instead of a
```

HPA ContainerResource type metric graduates to stable in v1.30. It allows you to configure
Autoscaling based on resource usage of individual containers. See [our previous
blog](/blog/2023/05/02/hpa-container-resource-metric/) for further details.

### CEL for admission control ([KEP-3488](https://kep.k8s.io/3488))

Integrating Common Expression Language (CEL) for admission control in Kubernetes introduces a more
dynamic and expressive way of evaluating admission requests. This feature allows complex,
fine-grained policies to be defined and enforced directly through the Kubernetes API, enhancing
security and governance capabilities without compromising performance or flexibility.

CEL's addition to Kubernetes admission control empowers cluster administrators to craft intricate
rules that can evaluate the content of API requests against the desired state and policies of the
cluster without resorting to Webhook-based access controllers. This level of control is crucial for
maintaining the integrity, security, and efficiency of cluster operations, making Kubernetes
environments more robust and adaptable to various use cases and requirements. For more information

> on the usage of CEL for admission control, see the [API

Reviewer 2: ```suggestion
on using CEL for admission control, see the [API
```

documentation](/docs/reference/access-authn-authz/validating-admission-policy/) for

> ValidatingAdmissionPolicy.

Author: > Perhaps incorporating something similar to what was done in the past mid-cycle blog ([here](https://kubernetes.io/blog/2023/11/16/kubernetes-1-29-upcoming-changes/)).

@dipesh-rawat We originally had a similar section, but it was suggested to remove it since this blog is not focused on deprecations and removals.

---
layout: blog

> title: 'Kubernetes Removals and Major Changes in Kubernetes 1.30'

Reviewer 3: ```suggestion
title: 'A Peek at Kubernetes v1.30'
```

date: 2024-03-08
slug: kubernetes-1-30-upcoming-changes
---

**Authors:** Amit Dsouza, Frederick Kautz, Kristin Martin, Abigail McCarthy, Natali Vlatko 

> 

Reviewer 3: ```suggestion

## A Quick Look: Exciting Changes in Kubernetes v1.30

It's a new year, and a new Kubernetes release. We're currently halfway through the release cycle and we have quite a few interesting and exciting enhancements coming with v1.30. From brand new features in alpha, to established features graduating to stable, to long-awaited improvements, this release has something for everyone to pay attention to!

To tide you over until the official release, here's a sneak peak of the enhancements we're most excited about this cycle!
```

​​Structured Parameters for dynamic resource allocation an extension to the original implementation that addresses this problem by building a framework to support making these claim parameters less opaque. Instead of handling the semantics of all claim parameters themselves, drivers could manage resources and describe them using a specific "structured model" pre-defined by Kubernetes. This would allow components aware of this "structured model" to make decisions about these resources without outsourcing them to some third-party controller. For example, the scheduler would be able to allocate claims rapidly, without back-and-forth communication with dynamic resource allocation drivers. Work done for this release centers on defining the framework necessary to enable different "structured models" and to implement the "named resources" model. This model allows listing individual resource instances and, compared to the traditional device plugin API, adds the ability to select those instances individually via attributes.

### Node memory swap support ([KEP-2400](https://kep.k8s.io/2400))

> In Kubernetes v1.30, node memory swap support graduates to `Beta2` with a strong emphasis on system stability. In previous Kubernetes versions, the `NodeSwap` feature gate was disabled by default and, when enabled, it used `UnlimitedSwap` behavior as the default behavior. To achieve better stability, `UnlimitedSwap` behavior which might compromise node stability will be removed in v1.30. In the Beta2 version, the `NodeSwap` is enabled by default with a new setting `NoSwap` being the default behavior. `NoSwap` turns off swap, giving administrators the opportunity to explicitly turn on swap for the desired nodes. Additional changes include updating implementation [best practices guidelines](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#best-practices) for `NodeSwap` with critical findings around keeping your system stable. For more details on node swap support in Kubernetes, read previous the [blog post](/blog/2023/08/24/swap-linux-beta/) or the [node swap documentation](/docs/concepts/architecture/nodes/#swap-memory).

Reviewer: However we explain this, we ought to make sure that readers understand that _just_ upgrading to v1.30 and turning swap space on is not enough, and is in fact likely to break your cluster if you only do that.

> In Kubernetes v1.30, node memory swap support graduates to `Beta2` with a strong emphasis on system stability. In previous Kubernetes versions, the `NodeSwap` feature gate was disabled by default and, when enabled, it used `UnlimitedSwap` behavior as the default behavior. To achieve better stability, `UnlimitedSwap` behavior which might compromise node stability will be removed in v1.30. In the Beta2 version, the `NodeSwap` is enabled by default with a new setting `NoSwap` being the default behavior. `NoSwap` turns off swap, giving administrators the opportunity to explicitly turn on swap for the desired nodes. Additional changes include updating implementation [best practices guidelines](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#best-practices) for `NodeSwap` with critical findings around keeping your system stable. For more details on node swap support in Kubernetes, read previous the [blog post](/blog/2023/08/24/swap-linux-beta/) or the [node swap documentation](/docs/concepts/architecture/nodes/#swap-memory).

Reviewer 4: Thanks for the suggestion @sftim. Your suggestion looks good to me. @iholder101, @kannon92 can you review and give feedback here

> In Kubernetes v1.30, node memory swap support graduates to `Beta2` with a strong emphasis on system stability. In previous Kubernetes versions, the `NodeSwap` feature gate was disabled by default and, when enabled, it used `UnlimitedSwap` behavior as the default behavior. To achieve better stability, `UnlimitedSwap` behavior which might compromise node stability will be removed in v1.30. In the Beta2 version, the `NodeSwap` is enabled by default with a new setting `NoSwap` being the default behavior. `NoSwap` turns off swap, giving administrators the opportunity to explicitly turn on swap for the desired nodes. Additional changes include updating implementation [best practices guidelines](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#best-practices) for `NodeSwap` with critical findings around keeping your system stable. For more details on node swap support in Kubernetes, read previous the [blog post](/blog/2023/08/24/swap-linux-beta/) or the [node swap documentation](/docs/concepts/architecture/nodes/#swap-memory).

Reviewer 5: LGTM!

> In Kubernetes v1.30, node memory swap support graduates to `Beta2` with a strong emphasis on system stability. In previous Kubernetes versions, the `NodeSwap` feature gate was disabled by default and, when enabled, it used `UnlimitedSwap` behavior as the default behavior. To achieve better stability, `UnlimitedSwap` behavior which might compromise node stability will be removed in v1.30. In the Beta2 version, the `NodeSwap` is enabled by default with a new setting `NoSwap` being the default behavior. `NoSwap` turns off swap, giving administrators the opportunity to explicitly turn on swap for the desired nodes. Additional changes include updating implementation [best practices guidelines](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#best-practices) for `NodeSwap` with critical findings around keeping your system stable. For more details on node swap support in Kubernetes, read previous the [blog post](/blog/2023/08/24/swap-linux-beta/) or the [node swap documentation](/docs/concepts/architecture/nodes/#swap-memory).

Reviewer 6: @sftim looks great! thank you!

One small note though - I wonder if we should still provide a link to the best practices documented in the KEP [1]. I think these best practices are important and perhaps we should advertise them here. WDYT?

[1] https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#best-practices

### Support User Namespaces in pods ([KEP-127](https://kep.k8s.io/127))

> [User namespaces](/docs/concepts/workloads/pods/user-namespaces) is a feature to better isolate pods to prevent or mitigate several CVEs rated high/critical, including the just published [CVE-2024-21626](https://github.com/opencontainers/runc/security/advisories/GHSA-xr7r-f8xq-vfvv). In Kubernetes 1.30, support for user namespaces is migrating to beta, and now supports pods with and without volumes, custom UID/GID ranges and more!

Reviewer: We should make it clear that this is for Linux nodes.

Reviewer 7: Like this, then?

```suggestion
[User namespaces](/docs/concepts/workloads/pods/user-namespaces) is a linux-only feature to better isolate pods to prevent or mitigate several CVEs rated high/critical, including the just published [CVE-2024-21626](https://github.com/opencontainers/runc/security/advisories/GHSA-xr7r-f8xq-vfvv). In Kubernetes 1.30, support for user namespaces is migrating to beta, and now supports pods with and without volumes, custom UID/GID ranges and more!
```

### Structured Authorization Configuration ([KEP-3221](https://kep.k8s.io/3221))

> The Kubernetes API server's authorizer chain can now be configured using a configuration file. This [feature](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3221-structured-authorization-configuration) enables the creation of authorization chains with multiple webhooks with well-defined parameters that validate requests in a particular order and allows fine-grained control – such as explicit Deny on failures. The configuration file allows one to specify CEL rules to pre-filter requests before they are dispatched to webhooks to prevent unnecessary invocations. API Server will also automatically reload the authorizer chain when the configuration file is modified.

Author: @natalisucks Can you sync with the KEP owner on these proposed changes?

> The Kubernetes API server's authorizer chain can now be configured using a configuration file. This [feature](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3221-structured-authorization-configuration) enables the creation of authorization chains with multiple webhooks with well-defined parameters that validate requests in a particular order and allows fine-grained control – such as explicit Deny on failures. The configuration file allows one to specify CEL rules to pre-filter requests before they are dispatched to webhooks to prevent unnecessary invocations. API Server will also automatically reload the authorizer chain when the configuration file is modified.

Reviewer 8: tagging @palnabarun and @ritazh to check on the above

> The Kubernetes API server's authorizer chain can now be configured using a configuration file. This [feature](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3221-structured-authorization-configuration) enables the creation of authorization chains with multiple webhooks with well-defined parameters that validate requests in a particular order and allows fine-grained control – such as explicit Deny on failures. The configuration file allows one to specify CEL rules to pre-filter requests before they are dispatched to webhooks to prevent unnecessary invocations. API Server will also automatically reload the authorizer chain when the configuration file is modified.

Reviewer 9: We should remove `complex" as simple use cases can still leverage the structured authz configuration feature.
Suggestion: 
> Structured Authorization Configuration feature is moving to beta and will be enable by default.

We should remove mentions of alpha to reduce confusion and focus on what users will get in beta.
Suggestion:
> This feature enables the creation of authorization chains with multiple webhooks with well-defined parameters that validate requests in a particular order and allows fine-grained control – such as explicit Deny on failures. The configuration file approach ...

An example configuration with all possible values is provided in the [Authorization docs](/docs/reference/access-authn-authz/authorization/#configuring-the-api-server-using-an-authorization-config-file). You must specify the path to that authorization configuration using the `--authorization-config` command line argument. While the feature is in the Alpha or Beta stage, there is no change if you want to keep on using command line flags instead of using a configuration file. From Kubernetes 1.30, the feature would be turned on by default. For more details, read the [Authorization docs](/docs/reference/access-authn-authz/authorization/#configuring-the-api-server-using-an-authorization-config-file).

> ### CEL for Admission Control ([KEP-3488](https://kep.k8s.io/3488))

Reviewer: This is a graduation to stable, so I'd list it either first or last.

Integrating Common Expression Language (CEL) for admission control in Kubernetes introduces a more dynamic and expressive way of evaluating admission requests. This feature allows complex, fine-grained policies to be defined and enforced directly through the Kubernetes API, enhancing security and governance capabilities without compromising performance or flexibility.

> CEL's addition to Kubernetes admission control empowers cluster administrators to craft intricate rules that can evaluate the content of API requests against the desired state and policies of the cluster without resorting to Webhook-based access controllers. This level of control is crucial for maintaining the integrity, security, and efficiency of cluster operations, making Kubernetes environments more robust and adaptable to various use cases and requirements.

Reviewer 10: +1 to a link to the API (https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/).

Looks really good to me otherwise.

> CEL's addition to Kubernetes admission control empowers cluster administrators to craft intricate rules that can evaluate the content of API requests against the desired state and policies of the cluster without resorting to Webhook-based access controllers. This level of control is crucial for maintaining the integrity, security, and efficiency of cluster operations, making Kubernetes environments more robust and adaptable to various use cases and requirements.

Reviewer 11: Similarly, is the CEL announcement that it has reached GA, rather than being `beta`?
Just when linking to the docs it looks like it has already been a beta feature since 1.28?

### Container Resource based Pod Autoscaling ([KEP-1610](https://kep.k8s.io/1610))

HPA ContainerResource type metric graduates to stable in v1.30. It allows us to configure Autoscaling based on resource usage of individual containers. 
See [our previous blog](/blog/2023/05/02/hpa-container-resource-metric/) for further details.

## Want to know more?

Deprecations are announced in the Kubernetes release notes. You can see the announcements of pending deprecations in the release notes for:

* [Kubernetes v1.26](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.26.md#deprecation)
* [Kubernetes v1.27](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.27.md#deprecation)
* [Kubernetes v1.28](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.28.md#deprecation)
* [Kubernetes v1.29](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.29.md#deprecation)

We will formally announce the deprecations that come with [Kubernetes v1.30](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.30.md#deprecation) as part of the CHANGELOG for that release.

> For information on the deprecation and removal process, refer to the official Kubernetes [deprecation policy](/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api) document.

Author: > Not sure if this fits the theme, what do folks think?

It should likely be removed since we're no longer discussing deprecations--I will check with the team.

To tide you over until the official release, here's a sneak peak of the enhancements we're most excited about this cycle!

## Major changes for Kubernetes v1.30

### Structured parameters for dynamic resource allocation ([KEP-4381](https://kep.k8s.io/4381))

[Dynamic resource allocation](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/) was added to Kubernetes as an alpha feature in v1.26. It defines an alternative to the traditional device-plugin API for requesting access to third-party resources. By design, dynamic resource allocation uses parameters for resources that are completely opaque to core Kubernetes. This approach poses a problem for the Cluster Autoscaler (CA) or for any higher level controller that needs to make decisions for a group of pods (e.g. a job scheduler). It cannot simulate the effect of allocating or deallocating claims over time. Only the third-party DRA drivers have the information available to do this.

> ​​Structured Parameters for dynamic resource allocation an extension to the original implementation that addresses this problem by building a framework to support making these claim parameters less opaque. Instead of handling the semantics of all claim parameters themselves, drivers could manage resources and describe them using a specific "structured model" pre-defined by Kubernetes. This would allow components aware of this "structured model" to make decisions about these resources without outsourcing them to some third-party controller. For example, the scheduler would be able to allocate claims rapidly, without back-and-forth communication with dynamic resource allocation drivers. Work done for this release centers on defining the framework necessary to enable different "structured models" and to implement the "named resources" model. This model allows listing individual resource instances and, compared to the traditional device plugin API, adds the ability to select those instances individually via attributes.

Reviewer 12: ```suggestion
​​Structured Parameters for dynamic resource allocation is an extension to the original implementation that addresses this problem by building a framework to support making these claim parameters less opaque. Instead of handling the semantics of all claim parameters themselves, drivers could manage resources and describe them using a specific "structured model" pre-defined by Kubernetes. This would allow components aware of this "structured model" to make decisions about these resources without outsourcing them to some third-party controller. For example, the scheduler would be able to allocate claims rapidly, without back-and-forth communication with dynamic resource allocation drivers. Work done for this release centers on defining the framework necessary to enable different "structured models" and to implement the "named resources" model. This model allows listing individual resource instances and, compared to the traditional device plugin API, adds the ability to select those instances individually via attributes.
```

The comment about line wrapping applies here.

Except for the missing "is" this looks good to me.

Now I just need to get https://github.com/kubernetes/kubernetes/pull/123516/ actually merged... exception request is pending.

### Structured authorization configuration ([KEP-3221](https://kep.k8s.io/3221))

The Kubernetes API server's authorizer chain can now be configured using a configuration file. This [feature](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3221-structured-authorization-configuration) enables the creation of authorization chains with multiple webhooks with well-defined parameters that validate requests in a particular order and allows fine-grained control – such as explicit Deny on failures. The configuration file allows one to specify CEL rules to pre-filter requests before they are dispatched to webhooks to prevent unnecessary invocations. API Server will also automatically reload the authorizer chain when the configuration file is modified.

> An example configuration with all possible values is provided in the [Authorization docs](/docs/reference/access-authn-authz/authorization/#configuring-the-api-server-using-an-authorization-config-file). You must specify the path to that authorization configuration using the `--authorization-config` command line argument. While the feature is in the Alpha or Beta stage, there is no change if you want to keep on using command line flags instead of using a configuration file. From Kubernetes 1.30, the feature would be turned on by default. For more details, read the [Authorization docs](/docs/reference/access-authn-authz/authorization/#configuring-the-api-server-using-an-authorization-config-file).

Reviewer 13: > While the feature is in the Alpha or Beta stage, there is no change if you want to keep on using command line flags instead of using a configuration file.

This is not limited to Alpha/Beta ... command-line flags for authorization configuration will continue to work as-is, or you can switch to putting options in an `--authorization-config` file to gain access to new authorization webhook capabilities like multiple webhooks, failure policy, and pre-filter rules.

> From Kubernetes 1.30, the feature would be turned on by default.

Prior to Kubernetes 1.30, the configuration file format is alpha-level and can be used by specifying `--authorization-config` and opting into the alpha feature with `--feature-gates=StructuredAuthenticationConfiguration=true`.

From Kubernetes 1.30, the configuration file format is beta-level and can be used by specifying `--authorization-config`.

Author: > > While the feature is in the Alpha or Beta stage, there is no change if you want to keep on using command line flags instead of using a configuration file.
> 
> This is not limited to Alpha/Beta ... command-line flags for authorization configuration will continue to work as-is, or you can switch to putting options in an `--authorization-config` file to gain access to new authorization webhook capabilities like multiple webhooks, failure policy, and pre-filter rules.
> 
> > From Kubernetes 1.30, the feature would be turned on by default.
> 
> Prior to Kubernetes 1.30, the configuration file format is alpha-level and can be used by specifying `--authorization-config` and opting into the alpha feature with `--feature-gates=StructuredAuthenticationConfiguration=true`.
> 
> From Kubernetes 1.30, the configuration file format is beta-level and can be used by specifying `--authorization-config`.

Can you please clarify the chang(s) you are suggesting here? @liggitt

Reviewer 13: Avoid making it sound like the current command line flags will only keep working during alpha / beta. I'd suggest something like this:

> If you want to keep on using command line flags instead of using a configuration file, those will continue to work as-is. To gain access to new authorization webhook capabilities like multiple webhooks, failure policy, and pre-filter rules, switch to putting options in an `--authorization-config` file. From Kubernetes 1.30, the configuration file format is beta-level and can be used by specifying `--authorization-config` since the feature gate is enabled by default.

### CEL for admission control ([KEP-3488](https://kep.k8s.io/3488))

Integrating Common Expression Language (CEL) for admission control in Kubernetes introduces a more dynamic and expressive way of evaluating admission requests. This feature allows complex, fine-grained policies to be defined and enforced directly through the Kubernetes API, enhancing security and governance capabilities without compromising performance or flexibility.

> CEL's addition to Kubernetes admission control empowers cluster administrators to craft intricate rules that can evaluate the content of API requests against the desired state and policies of the cluster without resorting to Webhook-based access controllers. This level of control is crucial for maintaining the integrity, security, and efficiency of cluster operations, making Kubernetes environments more robust and adaptable to various use cases and requirements.

Reviewer 3: ```suggestion
CEL's addition to Kubernetes admission control empowers cluster administrators to craft intricate rules that can evaluate the content of API requests against the desired state and policies of the cluster without resorting to Webhook-based access controllers. This level of control is crucial for maintaining the integrity, security, and efficiency of cluster operations, making Kubernetes environments more robust and adaptable to various use cases and requirements. For more information on the usage of CEL for admission control, see the [API documentation](/docs/reference/access-authn-authz/validating-admission-policy/) for ValidatingAdmissionPolicy.
```

​​Structured Parameters for dynamic resource allocation an extension to the original implementation that addresses this problem by building a framework to support making these claim parameters less opaque. Instead of handling the semantics of all claim parameters themselves, drivers could manage resources and describe them using a specific "structured model" pre-defined by Kubernetes. This would allow components aware of this "structured model" to make decisions about these resources without outsourcing them to some third-party controller. For example, the scheduler would be able to allocate claims rapidly, without back-and-forth communication with dynamic resource allocation drivers. Work done for this release centers on defining the framework necessary to enable different "structured models" and to implement the "named resources" model. This model allows listing individual resource instances and, compared to the traditional device plugin API, adds the ability to select those instances individually via attributes.

### Node memory swap support ([KEP-2400](https://kep.k8s.io/2400))

> In Kubernetes v1.30, node memory swap support graduates to `Beta2` with a strong emphasis on system stability. In previous Kubernetes versions, the `NodeSwap` feature gate was disabled by default and, when enabled, it used `UnlimitedSwap` behavior as the default behavior. To achieve better stability, `UnlimitedSwap` behavior which might compromise node stability will be removed in v1.30. In the Beta2 version, the `NodeSwap` is enabled by default with a new setting `NoSwap` being the default behavior. `NoSwap` turns off swap, giving administrators the opportunity to explicitly turn on swap for the desired nodes. Additional changes include updating implementation [best practices guidelines](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#best-practices) for `NodeSwap` with critical findings around keeping your system stable. For more details on node swap support in Kubernetes, read previous the [blog post](https://kubernetes.io/blog/2023/08/24/swap-linux-beta/) or the [node swap documentation](https://kubernetes.io/docs/concepts/architecture/nodes/#swap-memory).

Reviewer 14: Consider utilizing site-relative links for the two direct links.

```suggestion
In Kubernetes v1.30, node memory swap support graduates to `Beta2` with a strong emphasis on system stability. In previous Kubernetes versions, the `NodeSwap` feature gate was disabled by default and, when enabled, it used `UnlimitedSwap` behavior as the default behavior. To achieve better stability, `UnlimitedSwap` behavior which might compromise node stability will be removed in v1.30. In the Beta2 version, the `NodeSwap` is enabled by default with a new setting `NoSwap` being the default behavior. `NoSwap` turns off swap, giving administrators the opportunity to explicitly turn on swap for the desired nodes. Additional changes include updating implementation [best practices guidelines](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#best-practices) for `NodeSwap` with critical findings around keeping your system stable. For more details on node swap support in Kubernetes, read previous the [blog post](/blog/2023/08/24/swap-linux-beta/) or the [node swap documentation](/docs/concepts/architecture/nodes/#swap-memory).
```

### CEL for Admission Control ([KEP-3488](https://kep.k8s.io/3488))

Integrating Common Expression Language (CEL) for admission control in Kubernetes introduces a more dynamic and expressive way of evaluating admission requests. This feature allows complex, fine-grained policies to be defined and enforced directly through the Kubernetes API, enhancing security and governance capabilities without compromising performance or flexibility.

CEL's addition to Kubernetes admission control empowers cluster administrators to craft intricate rules that can evaluate the content of API requests against the desired state and policies of the cluster without resorting to Webhook-based access controllers. This level of control is crucial for maintaining the integrity, security, and efficiency of cluster operations, making Kubernetes environments more robust and adaptable to various use cases and requirements.

### Container Resource based Pod Autoscaling ([KEP-1610](https://kep.k8s.io/1610))

> HPA ContainerResource type metric graduates to stable in v1.30. It allows us to configure Autoscaling based on resource usage of individual containers. See [our previous blog](https://kubernetes.io/blog/2023/05/02/hpa-container-resource-metric/) for further details.

Reviewer 14: Consider using the site-relative link for the blog URL.

```suggestion
HPA ContainerResource type metric graduates to stable in v1.30. It allows us to configure Autoscaling based on resource usage of individual containers. 
See [our previous blog](/blog/2023/05/02/hpa-container-resource-metric/) for further details.
```

Author: Thank you! Are the site-relative links documented somewhere (we would like to add the usage guidelines for these to our handbook)?

* Generally available (GA) or stable API versions may be marked as deprecated, but must not be removed within a major version of Kubernetes.
* Beta or pre-release API versions must be supported for 3 releases after deprecation.
* Alpha or experimental API versions may be removed in any release without prior deprecation notice.

Whether an API is removed as a result of a feature graduating from beta to stable or because that API simply did not succeed, all removals comply with this deprecation policy. Whenever an API is removed, migration options are communicated in the documentation.

## Removal for Kubernetes v1.30

### Removal of `SecurityContextDeny` admission plugin

> The `SecurityContextDeny` admission plugin, deprecated since v1.27, is being [removed in v1.30](https://github.com/kubernetes/kubernetes/pull/122612). The `PodSecurity` admission plugin, [available since v1.25](/docs/concepts/security/pod-security-admission/), is recommended instead. See [Admission Controllers reference](/docs/reference/access-authn-authz/admission-controllers/#securitycontextdeny) for more information. 

Reviewer: How about dropping this, and focusing on the “upcoming changes” aspect only? The actual removal part is quite a trivial one.
